### ISSUE TITLE
Quarks: dependencies fail if already installed

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#1462]
When a Quark is already installed, Quarks with a dependency on the former will fail to install. Posting e.g.:

````
ERROR: A version of Quark: GamelanSynthesis is already installed at /Volumes/data/sc/quarks/GamelanSynthesis
ERROR: Failed to install VirtualGamelan
````

This may have to do with versions in different folders (development vs. supercollider quark repository).

**Comment by crucialfelix** [comment]
what is the exact sequence you are doing where you get this ?

    Quarks.install("VirtualGamelan");
    Quarks.install("GamelanSynthesis");

?

**Comment by telephon** [comment]
Ok, so there seem to be a number of issues. I am using my own development version of the Quarks, in an added folder, not in ````downloaded-quarks````. If I uninstall all related quarks, and install one with dependencies (e.g. Gamelan).

then it installs both versions, my developing one and the public one, and I get the notorious error:

````
ERROR: duplicate Class found: 'VGTuning' 
/Volumes/data/sc/quarks/GamelanSynthesis/Classes/VGTuning.sc
~/Library/Application Support/SuperCollider/downloaded-quarks/GamelanSynthesis/Classes/VGTuning.sc
````

Starting out clean, and then only installing one dependency (say, ListeningClocks) _before_ installing the  GamelanSynthesis quark, I get the error: ````Error while installing Quark: VirtualGamelan[0.1]```` in the GUI, and:

````
ERROR: A version of Quark: ListeningClocks[0.1] is already installed at /Volumes/data/sc/quarks/ListeningClocks
ERROR: Failed to install VirtualGamelan
````

in the commandline.

**Comment by crucialfelix** [comment]
So you have a dependency specified as merely a name "ListeningClocks" which means that the localPath is assumed to be in downloaded-quarks/ListeningClocks

so it should check all folders in additionalFolders.

but it also means a lot of other changes in order to support additionalFolders completely.

**Comment by crucialfelix** [comment]
I removed the bug tag because you've added additionalFolders and you have various conflicting copies. To fully support additionalFolders is an enhancement

**Comment by telephon** [comment]
It's ok for now. 

But of course this behavior will bite everyone who maintains their own quarks on their own repository rather than the supercollider-quarks one.

**Comment by crucialfelix** [comment]
It has nothing to do with the repository that the quark comes from. Quarks may come from any repository.

The issue is adding extra folders for working copies. Currently a Quark is equal if the localPaths are equal. Quark-isInstalled needs to search each of the folders for any folder matching its name.

**Comment by telephon** [comment]
>It has nothing to do with the repository that the quark comes from. Quarks may come from any repository.

Yes, but if you maintain your own repository, you have to have folder *other than downloaded-quarks*.

Well, I simply had supposed that it would search for dependencies first in its own directory and then elsewhere.

**Comment by crucialfelix** [comment]
You can install a quark from any repository.

Quarks.install("https://bitbucket.org/telon/private-thing")

If you maintain your own repository you use whatever url and scheme your
repository can support:

Quarks.install("git://123.45.67.8/private-thing")

They will all install into downloaded quarks.


On Fri, Jul 3, 2015 at 7:40 PM Julian Rohrhuber <notifications@github.com>
wrote:

> It has nothing to do with the repository that the quark comes from. Quarks
> may come from any repository.
>
> Yes, but if you maintain your own repository, you have to have folder *other
> than downloaded-quarks*.
>
Not at all.

Maybe you mean that you want to have a separate working copy that you will
edit and you don't want to do your work in downloaded-quarks.

You don't have to do that, but you may want to do that.


> Well, I simply had supposed that it would search for dependencies first in
> its own directory and then elsewhere.
>

yes, that's the issue that will need to be fixed.  but that has nothing to
do with where the repository is or whether its "yours" or not.

**Comment by telephon** [comment]
Good to know that ````downloaded-quarks```` that are from your own repo, which is in turn linked with the supercollider-quarks repo won't cause conflicts. I had assumed so.

But really, I tend to want a separate folder for them, just to be sure where I am (because otherwise it is not straightforward to tell where the quark has come from).
